From 536228fef7bc334c42c950ef06a749ee68060570 Mon Sep 17 00:00:00 2001
From: Vinita Gupta <vgupta@caviumnetworks.com>
Date: Fri, 13 Feb 2015 14:38:17 +0530
Subject: [PATCH 081/148] MIPS:OCTEON:SATA:hecked ata platform code to modify
 octeon SHIM according to endianess

Commit 8dd47ca4b86d7199777478777dfa5d4e591bb873 from
git://git.yoctoproject.org/linux-yocto-3.14

Add compatibility for Octeon

Signed-off-by: Vinita Gupta  <vgupta@caviumnetworks.com>
Signed-off-by: Abhishek Paliwal <abhishek.paliwal@aricent.com>
---
 drivers/ata/Kconfig         |  5 +++++
 drivers/ata/Makefile        |  1 +
 drivers/ata/ahci_platform.c |  5 +++++
 drivers/ata/sata_octeon.c   | 44 ++++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 55 insertions(+)
 create mode 100644 drivers/ata/sata_octeon.c

diff --git a/drivers/ata/Kconfig b/drivers/ata/Kconfig
index 868429a..b61071e 100644
--- a/drivers/ata/Kconfig
+++ b/drivers/ata/Kconfig
@@ -155,6 +155,11 @@ config ATA_SFF
 
 	  If unsure, say Y.
 
+config SATA_OCTEON
+	tristate
+	depends on SATA_AHCI_PLATFORM && CAVIUM_OCTEON_SOC
+	default y
+
 if ATA_SFF
 
 comment "SFF controllers with custom DMA interface"
diff --git a/drivers/ata/Makefile b/drivers/ata/Makefile
index 46518c6..fcaf000 100644
--- a/drivers/ata/Makefile
+++ b/drivers/ata/Makefile
@@ -10,6 +10,7 @@ obj-$(CONFIG_SATA_INIC162X)	+= sata_inic162x.o
 obj-$(CONFIG_SATA_SIL24)	+= sata_sil24.o
 obj-$(CONFIG_SATA_DWC)		+= sata_dwc_460ex.o
 obj-$(CONFIG_SATA_HIGHBANK)	+= sata_highbank.o libahci.o
+obj-$(CONFIG_SATA_OCTEON)	+= sata_octeon.o
 obj-$(CONFIG_AHCI_IMX)		+= ahci_imx.o
 
 # SFF w/ custom DMA
diff --git a/drivers/ata/ahci_platform.c b/drivers/ata/ahci_platform.c
index 4b231ba..62c73da 100644
--- a/drivers/ata/ahci_platform.c
+++ b/drivers/ata/ahci_platform.c
@@ -26,6 +26,7 @@
 #include "ahci.h"
 
 static void ahci_host_stop(struct ata_host *host);
+extern void ahci_octeon_config(struct platform_device *pdev);
 
 enum ahci_type {
 	AHCI,		/* standard platform ahci */
@@ -102,6 +103,9 @@ static int ahci_probe(struct platform_device *pdev)
 	int i;
 	int rc;
 
+#if IS_ENABLED(CONFIG_SATA_OCTEON)
+	ahci_octeon_config(pdev);
+#endif
 	mem = platform_get_resource(pdev, IORESOURCE_MEM, 0);
 	if (!mem) {
 		dev_err(dev, "no mmio space\n");
@@ -330,6 +334,7 @@ static const struct of_device_id ahci_of_match[] = {
 	{ .compatible = "snps,spear-ahci", },
 	{ .compatible = "snps,exynos5440-ahci", },
 	{ .compatible = "ibm,476gtr-ahci", },
+	{ .compatible = "cavium,octeon-7130-ahci", },
 	{},
 };
 MODULE_DEVICE_TABLE(of, ahci_of_match);
diff --git a/drivers/ata/sata_octeon.c b/drivers/ata/sata_octeon.c
new file mode 100644
index 0000000..f757414
--- /dev/null
+++ b/drivers/ata/sata_octeon.c
@@ -0,0 +1,44 @@
+/*
+ * SATA glue for Cavium Octeon III SOCs.
+ *
+ *
+ * This file is subject to the terms and conditions of the GNU General Public
+ * License.  See the file "COPYING" in the main directory of this archive
+ * for more details.
+ *
+ * Copyright (C) 2010 Cavium Networks
+ *
+ */
+
+#include <linux/module.h>
+#include <linux/dma-mapping.h>
+#include <linux/platform_device.h>
+
+#include <asm/octeon/octeon.h>
+#include <asm/octeon/cvmx-sata-defs.h>
+
+void ahci_octeon_config(struct platform_device *pdev)
+{
+	cvmx_sata_uctl_shim_cfg_t shim_cfg;
+
+	/* set-up endian mode */
+	shim_cfg.u64 = cvmx_read_csr(CVMX_SATA_UCTL_SHIM_CFG);
+#ifdef __BIG_ENDIAN
+	shim_cfg.s.dma_endian_mode = 1;
+	shim_cfg.s.csr_endian_mode = 1;
+#else
+	shim_cfg.s.dma_endian_mode = 0;
+	shim_cfg.s.csr_endian_mode = 0;
+#endif
+	shim_cfg.s.dma_read_cmd = 1; /* No allocate L2C */
+	cvmx_write_csr(CVMX_SATA_UCTL_SHIM_CFG, shim_cfg.u64);
+
+	/* Set a good dma_mask */
+	pdev->dev.coherent_dma_mask = DMA_BIT_MASK(64);
+	pdev->dev.dma_mask = &pdev->dev.coherent_dma_mask;
+}
+EXPORT_SYMBOL(ahci_octeon_config);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Cavium, Inc. <support@cavium.com>");
+MODULE_DESCRIPTION("Cavium Inc. sata config.");
-- 
1.8.2.1

